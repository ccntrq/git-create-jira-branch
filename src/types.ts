import * as Schema from "@effect/schema/Schema";
import { Brand, Data } from "effect";

export type JiraApiUrl = string & Brand.Brand<"JiraApiUrl">;
export const JiraApiUrl = Brand.nominal<JiraApiUrl>();
export type JiraPat = string & Brand.Brand<"JiraPat">;
export const JiraPat = Brand.nominal<JiraPat>();
export type JiraKeyPrefix = string & Brand.Brand<"JiraKeyPrefix">;
export const JiraKeyPrefix = Brand.nominal<JiraKeyPrefix>();

export const JiraIssuetypeSchema = Schema.struct({
  name: Schema.string,
});

export type JiraIssuetype = Schema.Schema.To<typeof JiraIssuetypeSchema>;

export const JiraIssueSchema = Schema.struct({
  key: Schema.string,
  fields: Schema.struct({
    summary: Schema.string,
    issuetype: JiraIssuetypeSchema,
  }),
});

export type JiraIssue = Schema.Schema.To<typeof JiraIssueSchema>;

export type GitCreateJiraBranchError =
  | AppConfigError
  | GitExecError
  | JiraApiError;

export declare namespace GitCreateJiraBranchError {
  export interface Proto {
    readonly _tag: string;
    readonly message: string;
  }

  export type ProvidedFields = "_tag";
}

const makeError =
  <E extends GitCreateJiraBranchError.Proto>(tag: E["_tag"]) =>
  (args: Omit<E, GitCreateJiraBranchError.ProvidedFields>): E =>
    Data.struct({
      _tag: tag,
      ...args,
    } as E);

export interface GitExecError extends GitCreateJiraBranchError.Proto {
  readonly _tag: "GitExecError";
}

export const GitExecError = makeError<GitExecError>("GitExecError");

export interface AppConfigError extends GitCreateJiraBranchError.Proto {
  readonly _tag: "AppConfigError";
}
export const AppConfigError = makeError<AppConfigError>("AppConfigError");

export interface JiraApiError extends GitCreateJiraBranchError.Proto {
  readonly _tag: "JiraApiError";
}
export const JiraApiError = makeError<JiraApiError>("JiraApiError");
