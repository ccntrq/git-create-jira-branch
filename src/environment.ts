import {Config, ConfigError, Context, Effect, Layer} from 'effect';
import * as Option from 'effect/Option';

import {
  AppConfig,
  AppConfigError,
  JiraApiToken,
  JiraApiUrl,
  JiraCloudAuth,
  JiraDataCenterAuth,
  JiraKeyPrefix,
  JiraPat,
  JiraUserEmail,
} from './types';

export interface Environment {
  readonly getEnv: () => Effect.Effect<never, AppConfigError, AppConfig>;
}

export const Environment = Context.Tag<Environment>();

export const EnvironmentLive = Layer.succeed(
  Environment,
  Environment.of({
    getEnv: () =>
      Effect.all(
        [
          Config.string('JIRA_API_URL'),
          Config.option(Config.string('JIRA_PAT')),
          Config.option(Config.string('JIRA_USER_EMAIL')),
          Config.option(Config.string('JIRA_API_TOKEN')),
          Config.option(Config.string('JIRA_KEY_PREFIX')),
        ],
        {concurrency: 'unbounded', mode: 'validate'},
      ).pipe(
        Effect.catchAll((maybeErrors) => {
          return Effect.fail(
            AppConfigError({
              message: Option.reduceCompact<ConfigError.ConfigError, string[]>(
                maybeErrors,
                [],
                (acc, e) =>
                  acc.concat(
                    ConfigError.reduceWithContext(
                      e,
                      void 0,
                      mkCollectConfigErrorMessagesReducer(),
                    ),
                  ),
              ).join('\n'),
            }),
          );
        }),
        Effect.flatMap(
          ([
            jiraApiUrl,
            jiraPat,
            jiraUserEmail,
            jiraApiToken,
            defaultJiraKeyPrefix,
          ]) =>
            Effect.gen(function* (_) {
              if (
                Option.isSome(jiraPat) &&
                (Option.isSome(jiraUserEmail) || Option.isSome(jiraApiToken))
              ) {
                return yield* _(
                  Effect.fail(
                    AppConfigError({
                      message:
                        'Please provide either JIRA_PAT (for Jira Data Center) or JIRA_API_TOKEN and JIRA_USER_EMAIL (for Jira Cloud), but not both',
                    }),
                  ),
                );
              }

              if (Option.isSome(jiraPat)) {
                return {
                  jiraAuth: JiraDataCenterAuth({
                    jiraPat: JiraPat(jiraPat.value),
                  }),
                  jiraApiUrl: JiraApiUrl(jiraApiUrl),
                  defaultJiraKeyPrefix: Option.map(
                    defaultJiraKeyPrefix,
                    JiraKeyPrefix,
                  ),
                };
              }

              if (Option.isSome(jiraApiToken) && Option.isSome(jiraUserEmail)) {
                return {
                  jiraAuth: JiraCloudAuth({
                    jiraApiToken: JiraApiToken(jiraApiToken.value),
                    jiraUserEmail: JiraUserEmail(jiraUserEmail.value),
                  }),
                  jiraApiUrl: JiraApiUrl(jiraApiUrl),
                  defaultJiraKeyPrefix: Option.map(
                    defaultJiraKeyPrefix,
                    JiraKeyPrefix,
                  ),
                };
              }

              return yield* _(
                Effect.fail(
                  AppConfigError({
                    message:
                      'Please provide JIRA_PAT (for Jira Data Center) or JIRA_API_TOKEN and JIRA_USER_EMAIL (for Jira Cloud).',
                  }),
                ),
              );
            }),
        ),
      ),
  }),
);

const mkCollectConfigErrorMessagesReducer = (): ConfigError.ConfigErrorReducer<
  unknown,
  string[]
> => {
  const collectMessage = (
    _: unknown,
    _path: string[],
    message: string,
  ): string[] => [message];

  const mergeErrors = (
    _: unknown,
    left: string[],
    right: string[],
  ): string[] => [...left, ...right];

  return {
    andCase: mergeErrors,
    orCase: mergeErrors,
    invalidDataCase: collectMessage,
    missingDataCase: collectMessage,
    sourceUnavailableCase: collectMessage,
    unsupportedCase: collectMessage,
  };
};
