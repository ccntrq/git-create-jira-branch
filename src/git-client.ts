import { exec } from "child_process";
import { Context, Effect, Layer } from "effect";
import { GitExecError } from "./types";

type GitClientEffect<A> = Effect.Effect<never, GitExecError, A>;

export interface GitClient {
  readonly createGitBranch: (branchName: string) => GitClientEffect<void>;
  readonly createGitBranchFrom: (
    baseBranch: string
  ) => (branchName: string) => GitClientEffect<void>;
}

export const GitClient = Context.Tag<GitClient>();

const createGitBranchFrom =
  (baseBranch: string) =>
  (branchName: string): GitClientEffect<void> =>
    Effect.async((cb) => {
      exec(`git checkout -b ${branchName} ${baseBranch}`, (err) =>
        err !== null
          ? cb(
              Effect.fail(
                GitExecError({
                  message: `Checkout failed:\n'${err.message.trim()}'`,
                })
              )
            )
          : cb(Effect.succeed(undefined))
      );
    });

const gitClient = {
  createGitBranchFrom: createGitBranchFrom,
  createGitBranch: createGitBranchFrom(""),
};

export const GitClientLive = Layer.succeed(GitClient, GitClient.of(gitClient));
