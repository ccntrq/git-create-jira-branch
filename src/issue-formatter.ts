import type {JiraIssue} from './types';

import {Doc} from '@effect/printer';
import {bold, underlined} from '@effect/printer-ansi/Ansi';
import {render} from '@effect/printer-ansi/AnsiDoc';
import {pipe} from 'effect';

export const formatIssue = (issue: JiraIssue): string => {
  const heading = pipe(
    [
      mkBold(Doc.text(issue.key)),
      //Doc.text(issue.key),
      Doc.char('-'),
      Doc.text(issue.fields.summary),
    ],
    unwords,
    underline,
  );

  const details = pipe(
    [
      mkBold(Doc.text(issue.fields.issuetype.name)),
      Doc.char('|'),
      mkBold(Doc.text('Status:')),
      Doc.text(issue.fields.status.name),
      Doc.char('|'),
      mkBold(Doc.text('Creator:')),
      Doc.text(issue.fields.creator.displayName),
      ...(issue.fields.assignee
        ? [
            Doc.char('|'),
            mkBold(Doc.text('Assignee:')),
            Doc.text(issue.fields.assignee.displayName),
          ]
        : []),
    ],
    unwords,
    underline,
  );

  const description = issue.fields.description
    ? pipe(
        issue.fields.description.split(/\n|\r\n/),
        (d) =>
          // TODO:
          // use fillSep/reflow here to break lines at 80 characters when
          // performance issue is resolved
          // https://github.com/Effect-TS/effect/issues/2371
          // https://discord.com/channels/795981131316985866/1219736390401786008
          d.map(Doc.text),
        Doc.vcat,
      )
    : Doc.empty;

  const doc = pipe(
    [
      heading,
      details,
      ...(Doc.isEmpty(description) ? [] : [Doc.empty, description]),
    ],
    unlines,
  );

  return render(doc, {style: 'pretty'});
};

const unwords = <T>(docs: Iterable<Doc.Doc<T>>): Doc.Doc<T> =>
  Doc.concatWith<T>(Doc.catWithSpace)(docs);
const unlines = <T>(docs: Iterable<Doc.Doc<T>>): Doc.Doc<T> =>
  Doc.concatWith<T>(Doc.catWithLine)(docs);
const mkBold = Doc.annotate(bold);
const underline = Doc.annotate(underlined);
