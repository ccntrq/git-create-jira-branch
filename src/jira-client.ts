import { Layer, Context, Effect, pipe } from "effect";
import * as Http from "@effect/platform/HttpClient";
import * as ArrayFormatter from "@effect/schema/ArrayFormatter";
import * as ParseResult from "@effect/schema/ParseResult";

import { Environment } from "./environment";
import {
  AppConfigError,
  JiraApiError,
  JiraIssue,
  JiraIssueSchema,
} from "./types";

export interface JiraClient {
  readonly getJiraIssue: (
    issueKey: string
  ) => Effect.Effect<Environment, AppConfigError | JiraApiError, JiraIssue>;
}

export const JiraClient = Context.Tag<JiraClient>();

export const JiraClientLive = Layer.effect(
  JiraClient,
  Environment.pipe(
    Effect.map((env) =>
      JiraClient.of({
        getJiraIssue: (issueId: string) => {
          const endPoint = `/rest/api/latest/issue/${issueId}`;

          return env.getEnv().pipe(
            Effect.flatMap(({ jiraPat, jiraApiUrl }) =>
              Http.request
                .get(jiraApiUrl + endPoint, {
                  headers: {
                    Authorization: `Bearer ${jiraPat}`,
                    Accept: "application/json",
                  },
                })
                .pipe(Http.client.fetchOk())
            ),
            Effect.flatMap(Http.response.schemaBodyJson(JiraIssueSchema)),
            handleJiraClientErrors
          );
        },
      })
    )
  )
);

const handleJiraClientErrors = (
  eff: Effect.Effect<
    Environment,
    AppConfigError | ParseResult.ParseError | Http.error.HttpClientError,
    JiraIssue
  >
): Effect.Effect<Environment, AppConfigError | JiraApiError, JiraIssue> =>
  eff.pipe(
    Effect.catchTag("ParseError", (e) =>
      Effect.fail(
        JiraApiError({
          message: [
            "Failed to parse ticket response from Jira:",
            ...ArrayFormatter.formatErrors(e.errors).map(
              (issue) => `'${issue.path.join(" ")}': '${issue.message}'`
            ),
          ].join("\n"),
        })
      )
    ),
    Effect.catchTag("RequestError", (e) =>
      Effect.fail(
        JiraApiError({
          message: `Failed to make ticket request to Jira. Reason: ${e.reason}`,
        })
      )
    ),
    Effect.catchTag("ResponseError", (e) =>
      Effect.fail(
        JiraApiError({
          message: `Jira Ticket request returned failure. Reason: ${e.reason}${
            typeof e.error === "string" ? ` (${e.error})` : ""
          } StatusCode: ${e.response.status}`,
        })
      )
    )
  );
