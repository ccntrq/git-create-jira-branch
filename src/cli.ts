import {Option, Data, Effect, pipe, Console} from 'effect';
import {ValidationError} from '@effect/cli';
import {CliApp} from '@effect/cli/CliApp';
import {isNone} from 'effect/Option';
import * as Args from '@effect/cli/Args';
import * as Command from '@effect/cli/Command';
import * as Options from '@effect/cli/Options';
import * as Span from '@effect/cli/HelpDoc/Span';
import * as HelpDoc from '@effect/cli/HelpDoc';
import * as CommandExecutor from '@effect/platform/CommandExecutor';

import {
  gitCreateJiraBranch,
  ticketUrl,
  ticketUrlForCurrentBranch,
} from './core';
import * as packageJson from '../package.json';
import {
  GitCreateJiraBranchError,
  matchGitCreateJiraBranchResult,
} from './types';
import {compose} from 'effect/Function';
import {AppConfigService} from './app-config';
import {GitClient} from './git-client';
import {JiraClient} from './jira-client';
import {openUrl} from './url-opener';

interface GitCreateJiraBranch {
  readonly jiraKey: Option.Option<string>;
  readonly baseBranch: Option.Option<string>;
  readonly reset: boolean;
  readonly open: boolean;
}

const GitCreateJiraBranch = Data.case<GitCreateJiraBranch>();

const mainCommand = pipe(
  Command.make(
    'git-create-jira-branch',
    {
      options: Options.all({
        baseBranch: Options.withDescription(
          Options.optional(Options.withAlias(Options.text('baseBranch'), 'b')),
          'Base branch to create the new branch from',
        ),
        reset: Options.withDescription(
          Options.withAlias(Options.boolean('reset'), 'r'),
          'Reset the branch if it already exists',
        ),
        open: Options.withDescription(
          Options.withAlias(Options.boolean('open'), 'o'),
          'Open the ticket for the current branch or the given ticket in the browser',
        ),
      }),
      args: Args.withDescription(
        Args.atMost(Args.text({name: 'jira-key'}), 1),
        'The Jira ticket key to create a branch for (e.g. FOOX-1234)',
      ),
    },
    (
      args,
    ): Effect.Effect<
      void,
      GitCreateJiraBranchError,
      | AppConfigService
      | GitClient
      | JiraClient
      | CommandExecutor.CommandExecutor
    > => {
      const command = GitCreateJiraBranch({
        baseBranch: args.options.baseBranch,
        reset: args.options.reset,
        open: args.options.open,
        jiraKey: Option.fromNullable(args.args[0]),
      });

      if (command.open) {
        return pipe(
          command.jiraKey,
          Option.match({
            onSome: (jiraKey) => ticketUrl(jiraKey),
            onNone: () => ticketUrlForCurrentBranch(),
          }),
          Effect.tap((url) =>
            Console.log(
              `Opening ticket url '${url}' in your default browser...`,
            ),
          ),
          Effect.flatMap(openUrl),
        );
      }

      if (isNone(command.jiraKey)) {
        return printDocs(HelpDoc.p(Span.error('No Jira Key provided')));
      }

      return Effect.flatMap(
        gitCreateJiraBranch(
          command.jiraKey.value,
          command.baseBranch,
          command.reset,
        ),
        compose(
          matchGitCreateJiraBranchResult({
            onCreatedBranch: (branch) =>
              `Successfully created branch: '${branch}'`,
            onSwitchedBranch: (branch) =>
              `Switched to already existing branch: '${branch}'`,
            onResetBranch: (branch) => `Reset branch: '${branch}'`,
          }),
          Console.log,
        ),
      );
    },
  ),
  Command.withDescription(
    HelpDoc.p(
      `Fetches the given Jira ticket and creates an aproriately named branch for it.
The branch type (bug or feat) is determined by the ticket type. The branch name
is based on the ticket summary.`,
    ),
  ),
);

const cli = {
  name: packageJson.name,
  version: packageJson.version,
  summary: Span.text(packageJson.description),
};

export const cliEffect = (
  args: string[],
): Effect.Effect<
  void,
  never,
  | CliApp.Environment
  | GitClient
  | AppConfigService
  | JiraClient
  | CommandExecutor.CommandExecutor
> =>
  Command.run(
    mainCommand,
    cli,
  )(args).pipe(
    Effect.catchIf(ValidationError.isValidationError, (e) =>
      printDocs(HelpDoc.p(Span.error(HelpDoc.getSpan(e.error)))),
    ),
    Effect.catchAll((e) => printDocs(HelpDoc.p(Span.error(e.message)))),
  );

const printDocs = (doc: HelpDoc.HelpDoc): Effect.Effect<void> =>
  Console.log(HelpDoc.toAnsiText(doc));
