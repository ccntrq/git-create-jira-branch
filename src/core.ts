import {Chunk, Effect, Option} from 'effect';
import {isNone} from 'effect/Option';

import {GitClient} from './git-client';
import {JiraClient} from './jira-client';
import {Environment} from './environment';
import {
  CreatedBranch,
  GitCreateJiraBranchError,
  GitCreateJiraBranchResult,
  JiraIssue,
  JiraIssuetype,
  JiraKeyPrefix,
  SwitchedBranch,
} from './types';

export const gitCreateJiraBranch = (
  jiraKey: string,
  baseBranch: Option.Option<string>,
): Effect.Effect<
  Environment | GitClient | JiraClient,
  GitCreateJiraBranchError,
  GitCreateJiraBranchResult
> =>
  Effect.gen(function* ($) {
    const [envProvider, gitClient, jiraClient] = yield* $(
      Effect.all([Environment, GitClient, JiraClient]),
    );

    const {defaultJiraKeyPrefix} = yield* $(envProvider.getEnv());
    const fullKey = buildJiraKey(defaultJiraKeyPrefix, jiraKey);

    const issue = yield* $(jiraClient.getJiraIssue(fullKey));
    const branchName = jiraIssueToBranchName(issue);

    const branchExists = yield* $(
      Effect.map(gitClient.listBranches(), Chunk.contains(branchName)),
    );

    if (branchExists) {
      yield* $(gitClient.switchBranch(branchName));
      return SwitchedBranch(branchName);
    }

    yield* $(
      Option.match(baseBranch, {
        onNone: () => gitClient.createGitBranch,
        onSome: gitClient.createGitBranchFrom,
      })(branchName),
    );

    return CreatedBranch(branchName);
  });

const slugify = (str: string): string =>
  str
    .toLowerCase()
    .replace(/ä/g, 'ae')
    .replace(/ö/g, 'oe')
    .replace(/ü/g, 'ue')
    .replace(/ß/g, 'ss')
    .replace(/[^a-zA-Z0-9]/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '');

const jiraIssueToBranchName = (issue: JiraIssue): string => {
  const branchtype = jiraIssuetypeBranchtype(issue.fields.issuetype);
  return `${branchtype}/${issue.key}-${slugify(issue.fields.summary)}`;
};

const jiraIssuetypeBranchtype = (issuetype: JiraIssuetype): string => {
  if (issuetype.name.match(/bug/i)) {
    return 'fix';
  }

  return 'feat';
};

const buildJiraKey = (
  defaultJiraKeyPrefix: Option.Option<JiraKeyPrefix>,
  jiraKey: string,
): string =>
  jiraKey.match(/^([a-z]+)-(\d+)$/i) || isNone(defaultJiraKeyPrefix)
    ? jiraKey
    : `${defaultJiraKeyPrefix.value}-${jiraKey}`;
