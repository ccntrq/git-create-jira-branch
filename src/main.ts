#!/usr/bin/env node

import * as NodeCommandExecutor from '@effect/platform-node/NodeCommandExecutor';
import * as NodeFileSystem from '@effect/platform-node/NodeFileSystem';
import * as Http from '@effect/platform/HttpClient';
import {Cause, Console, Effect, Layer, pipe} from 'effect';

import {NodeContext} from '@effect/platform-node';
import {AppConfigService} from './app-config';
import {cliEffect} from './cli';
import {GitClientLive} from './git-client';
import {JiraClientLive} from './jira-client';

const commandExecutorLayer = NodeCommandExecutor.layer.pipe(
  Layer.provide(NodeFileSystem.layer),
);

const gitClientLayer = GitClientLive.pipe(Layer.provide(commandExecutorLayer));

const jiraClientLayer = JiraClientLive.pipe(
  Layer.provide(Http.client.layer),
  Layer.provide(AppConfigService.Live),
);

const mainLive = Layer.mergeAll(
  commandExecutorLayer,
  gitClientLayer,
  AppConfigService.Live,
  jiraClientLayer,
  NodeContext.layer,
);

const mainEffect = pipe(
  Effect.sync(() => process.argv),
  Effect.flatMap(cliEffect),
  Effect.provide(mainLive),
  Effect.catchAllDefect((defect) => {
    if (Cause.isRuntimeException(defect)) {
      return Console.log(`RuntimeException defect caught: ${defect.message}`);
    }
    return Console.log(`Unknown defect caught: ${JSON.stringify(defect)}`);
  }),
);

Effect.runPromise(mainEffect);
