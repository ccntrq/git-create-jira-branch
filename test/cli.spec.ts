import { Effect, Layer, pipe, Option } from "effect";
import { log } from "effect/Console";

import { Environment } from "../src/environment";
import { GitClient } from "../src/git-client";
import { JiraClient } from "../src/jira-client";
import { cliEffect } from "../src/cli";
import { gitCreateJiraBranch } from "../src/core";

import { vi, describe, afterEach, expect, beforeEach } from "vitest";
import { itEffect, toEffectMock } from "./util";

const mockGitClient = {
  createGitBranch: vi.fn(),
  createGitBranchFrom: vi.fn(),
};
const mockEnvironment = { getEnv: vi.fn() };
const mockJiraClient = { getJiraIssue: vi.fn() };

const testLayer = Layer.mergeAll(
  Layer.succeed(GitClient, GitClient.of(mockGitClient)),
  Layer.succeed(Environment, Environment.of(mockEnvironment)),
  Layer.succeed(JiraClient, JiraClient.of(mockJiraClient))
);

vi.mock("../src/core")

vi.mock("effect/Console");

const mockGitCreateJiraBranch = toEffectMock(gitCreateJiraBranch as any);
const mockLog = toEffectMock(log as any);

describe("cli", () => {
  describe("cliEffect", () => {
    afterEach(() => {
      vi.restoreAllMocks();
      mockLog.mockSuccessValue(undefined);
      mockGitCreateJiraBranch.mockSuccessValue("XXX/NO-BRANCH");
    });

    itEffect("should create branch with single argument", () =>
      Effect.gen(function* ($) {
        mockGitCreateJiraBranch.mockSuccessValue(
          "feat/FOOX-1234-description"
        );
        yield* $(
          Effect.provide(
            pipe(
              Effect.sync(() => ["FOOX-1234"]),
              Effect.flatMap(cliEffect)
            ),
            testLayer
          )
        );

        expect(mockGitCreateJiraBranch).toHaveBeenCalledWith(
          "FOOX-1234",
          Option.none()
        );
        expect(mockLog.mock.calls).toMatchSnapshot();
      })
    );

    itEffect("should handle basebranch argument (-b)", () =>
      Effect.gen(function* ($) {
        mockGitCreateJiraBranch.mockSuccessValue(
          "feat/FOOX-1234-description"
        );
        yield* $(
          Effect.provide(
            pipe(
              Effect.sync(() => ["FOOX-1234", "-b", "master"]),
              Effect.flatMap(cliEffect)
            ),
            testLayer
          )
        );

        expect(mockGitCreateJiraBranch).toHaveBeenCalledWith(
          "FOOX-1234",
          Option.some("master")
        );
        expect(mockLog.mock.calls).toMatchSnapshot();
      })
    );

    itEffect("should report missing jirakey", () =>
      Effect.gen(function* ($) {
        mockGitCreateJiraBranch.mockSuccessValue(
          "feat/FOOX-1234-description"
        );
        yield* $(
          Effect.provide(
            pipe(
              Effect.sync(() => []),
              Effect.flatMap(cliEffect)
            ),
            testLayer
          )
        );

        expect(mockGitCreateJiraBranch).not.toHaveBeenCalled();
        expect(mockLog.mock.calls).toMatchSnapshot();
      })
    );

    itEffect("should print help (--help)", () =>
      Effect.gen(function* ($) {
        mockGitCreateJiraBranch.mockSuccessValue(
          "feat/FOOX-1234-description"
        );
        yield* $(
          Effect.provide(
            pipe(
              Effect.sync(() => ["--help"]),
              Effect.flatMap(cliEffect)
            ),
            testLayer
          )
        );

        expect(mockGitCreateJiraBranch).not.toHaveBeenCalled();
        expect(mockLog.mock.calls).toMatchSnapshot();
      })
    );
  });
});
