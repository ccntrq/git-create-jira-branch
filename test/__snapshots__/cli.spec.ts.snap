// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cli > cliEffect > should create branch with single argument 1`] = `
[
  [
    "Successfully created branch: 'feat/FOOX-1234-description'",
  ],
]
`;

exports[`cli > cliEffect > should handle basebranch argument (-b) 1`] = `
[
  [
    "Successfully created branch: 'feat/FOOX-1234-description'",
  ],
]
`;

exports[`cli > cliEffect > should print help (--help) 1`] = `
[
  [
    "[0;1m[0;97;1mgit-create-jira-branch[0;1m[0m

git-create-jira-branch v0.1.0 -- Create a git branch from a Jira ticket

[0;1mUSAGE[0m

$ git-create-jira-branch [(-b, --baseBranch text)] [(-v, --version)] [(-h, --help)] <jira-key>...

[0;1mARGUMENTS[0m

[0;1m[0;30;1m<jira-key>[0;1m 0 - 1[0m

  A user-defined piece of text.

  The Jira ticket key to create a branch for (e.g. FOOX-1234)

  This argument must be repeated at least 0 times and may be repeated up to 1 times.

[0;1mOPTIONS[0m

[0;1m(-b, --baseBranch text)[0m

  A user-defined piece of text.

  Base branch to create the new branch from

  This setting is optional.

[0;1m(-v, --version)[0m

  A true or false value.

  Show version information

  This setting is optional. (Defaults to: 'false')

[0;1m(-h, --help)[0m

  A true or false value.

  Show this help text

  This setting is optional. (Defaults to: 'false')
",
  ],
]
`;

exports[`cli > cliEffect > should report missing jirakey 1`] = `
[
  [
    "[0;91mNo Jira Key provided[0m
",
  ],
]
`;
